packages/client/ui/react-ui/src/hooks/useWalletCache.ts

```diff
@@ -0,0 +1,132 @@
+import { useCallback } from "react";
+import type { Address } from "viem";
+import type { WebAuthnP256 } from "ox";
+import type { PasskeySigner } from "@crossmint/client-sdk-smart-wallet";
+
+type WalletCache = {
+    wallet?: {
+        address: Address;
+        initialized?: boolean;
+    };
+    passkey?: {
+        authenticatorId: string;
+        domain: string;
+        passkeyName: string;
+        pubKeyPrefix: number;
+        pubKeyX: string;
+        pubKeyY: string;
+    };
+};
+
+export function useWalletCache(userId: string | undefined) {
+    const getCache = useCallback((): WalletCache | null => {
+        if (userId == undefined) {
+            return null;
+        }
+        const key = `smart-wallet-${userId}`;
+        const cache = localStorage.getItem(key);
+        if (cache == null) {
+            return null;
+        }
+        return JSON.parse(cache) as WalletCache;
+    }, [userId]);
+
+    const setCache = useCallback(
+        (newCache: WalletCache) => {
+            if (userId == undefined) {
+                return;
+            }
+            const key = `smart-wallet-${userId}`;
+            localStorage.setItem(key, JSON.stringify(newCache));
+        },
+        [userId]
+    );
+
+    const getWalletAddress = useCallback((): Address | undefined => {
+        return getCache()?.wallet?.address;
+    }, [getCache]);
+
+    const setWalletAddress = useCallback(
+        (address: Address) => {
+            const currentCache = getCache() || {};
+            setCache({
+                ...currentCache,
+                wallet: {
+                    ...currentCache.wallet,
+                    address,
+                },
+            });
+        },
+        [getCache, setCache]
+    );
+
+    const isWalletInitialized = useCallback((): boolean => {
+        const cache = getCache();
+        if (cache == null) {
+            return false;
+        }
+        return cache.wallet?.initialized ?? false;
+    }, [getCache]);
+
+    const setWalletInitialized = useCallback(
+        (initialized: boolean) => {
+            const currentCache = getCache();
+            if (currentCache === null || currentCache.wallet === undefined) {
+                return;
+            }
+            setCache({
+                ...currentCache,
+                wallet: {
+                    ...currentCache.wallet,
+                    initialized,
+                },
+            });
+        },
+        [getCache, setCache]
+    );
+
+    const getPasskey = useCallback((): PasskeySigner | undefined => {
```

Comment:
why does this need to be a function instead of a state?

Reply:
Just a minor comment btw, I was thinking the sets could also save the state to both localstorage and an in memory state, but could be a silly optimization. It would make the code slightly cleaner as you could just fetch the values as is without getters. But leave it up to you

Reply:
you're right about function vs variable, rewrote with `useMemo` in https://github.com/Crossmint/crossmint-sdk/pull/971/commits/a1cd119ef6235485d7463dfe164784810816aab5

Reply:
love this!

Reply:
Not sure about merging localstorage and memory state. There's already `useWallet`, so changing `useWalletCache` to `useWalletState` or something like that would be confusing IMO. It's possible to do a bigger rewrite, but that will take more time and can introduce more bugs.

Probably doesn't make sense since this package will be deprecated/rewritten anyway.

Reply:
yeah, useMemo is more than enough!